### PR TITLE
Add tee command to Userland

### PR DESCRIPTION
**Pull request by vger92**
An initial stab at closing #109, would appreciate any feedback!

### CODE REVIEW

File: Userland/tee.cpp [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        if (nread == 0)
+            break;
+        if (nread < 0) {
+            perror("read() error");
+            *err = true;
+        }
+
+        for (int fd : fds) {
+            ssize_t nwrite = write(fd, buf, nread);
+            if (nwrite != nread)
```
Comment (awesomekling) [inline Userland/tee.cpp]: It's okay for `write()` to do a partial write, we shouldn't consider that an error, but instead retry until we've written everything. `write()` may also fail with `EINTR`, in which case I think we should also retry.

For other error codes, maybe we should print an error message to `stderr`?
And shouldn't we return on error?

File: Userland/tee.cpp [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+static void copy_stdin(Vector<int>& fds, bool* err)
+{
+    for (;;) {
+        char buf[4096];
+        ssize_t nread = read(STDIN_FILENO, buf, sizeof(buf));
+        if (nread == 0)
+            break;
+        if (nread < 0) {
+            perror("read() error");
+            *err = true;
```
Comment (awesomekling) [inline Userland/tee.cpp]: Shouldn't we return here?

File: Userland/tee.cpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    Vector<int> fds;
+    for (int i = start; i < argc; ++i) {
+        int fd = open(argv[i], oflag, mode);
+        if (fd < 0) {
+            perror("failed to open file for writing");
+            *err = true;
+        }
+        fds.append(fd);
```
Comment (docemh) [inline Userland/tee.cpp]: Perhaps
```suggestion
        else {
            fds.append(fd);
        }
```

File: Userland/tee.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+        oflag = O_CREAT | O_WRONLY | O_TRUNC;
+        mode = S_IROTH | S_IWOTH | S_IRGRP | S_IWGRP | S_IRUSR | S_IWUSR;
+    }
+
+    Vector<int> fds;
+    for (int i = start; i < argc; ++i) {
+        int fd = open(argv[i], oflag, mode);
+        if (fd < 0) {
+            perror("failed to open file for writing");
+            *err = true;
```
Comment (awesomekling) [inline Userland/tee.cpp]: Shouldn't we return here?

File: Userland/tee.cpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    if (iflag) {
+        if (signal(SIGINT, int_handler) == SIG_ERR) {
+            perror("failed to install SIGINT handler");
+        }
+    }
+
+    bool err_open = false, err_write = false;
+    auto fds = collect_fds(argc, argv, optind, aflag, &err_open);
```
Comment (awesomekling) [inline Userland/tee.cpp]: If `fds` is empty, is there really any point in calling `copy_stdin()`? It seems like we should bail at that point.

File: Userland/tee.cpp [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+        }
+    }
+
+    if (iflag) {
+        if (signal(SIGINT, int_handler) == SIG_ERR) {
+            perror("failed to install SIGINT handler");
+        }
+    }
+
+    bool err_open = false, err_write = false;
```
Comment (awesomekling) [inline Userland/tee.cpp]: Please put variable declarations on separate lines.

File: Userland/tee.cpp (hunk 3 above)
Comment (awesomekling) [reply Userland/tee.cpp]: Indeed. Any `write()` with an invalid fd argument will fail with `EINVAL` so there's no point in trying.

File: Userland/tee.cpp (hunk 4 above)
Comment (docemh) [reply Userland/tee.cpp]: According to my reading of the opengroup's documentation, writes should continue for all successfully opened files and stdout. So it makes sense to me to keep trying to open the remaining specified files.

File: Userland/tee.cpp (hunk 6 above)
Comment (docemh) [reply Userland/tee.cpp]: Sure

File: Userland/tee.cpp (hunk 4 above)
Comment (awesomekling) [reply Userland/tee.cpp]: Ah, right you are :)

File: Userland/tee.cpp (hunk 5 above)
Comment (docemh) [reply Userland/tee.cpp]: I don't think it's possible for `fds` to be empty, because `STDOUT_FILENO` is added to the vector unconditionally.

Comment (awesomekling) [reply Userland/tee.cpp]: Yeah, you're right. I was thinking `collect_fds()` could return without including `STDOUT_FILENO` but that's not accurate.

File: Userland/tee.cpp (hunk 2 above)
Comment (docemh) [reply Userland/tee.cpp]: Probably, but I wasn't able to find a specific section in the docs speaking to this behavior. If we fail to read from stdin though, things are probably done.

Comment (awesomekling) [reply Userland/tee.cpp]: How would we break out of this loop otherwise?

File: Userland/tee.cpp (hunk 1 above)
Comment (docemh) [reply Userland/tee.cpp]: Yes, sorry! partial writes should be allowed. `EINTR` will have to be handled, yes. The other error codes should have an error message, I agree. According to my understanding of the docs, we should stop trying to write to this current file, yes, but we should continue writing the other successfully opened files.

File: Userland/tee.cpp (hunk 2 above)
Comment (docemh) [reply Userland/tee.cpp]: Yes agreed

**Comment by docemh** [comment]
Second commit added to address these review comments. In summary: error handling, partial writes, and a couple style changes.

**Comment by awesomekling** [comment]
Oh, I just realized: when `write()` fails for an fd, we should remove that fd from `fds` so we don't keep trying to write to it.

**Comment by docemh** [comment]
Yes, agreed. Thanks for catching that.

**Pull request closed without merging**